scripts/update_robots.sh

```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
+
+# This file is part of solidity.
+#
+# solidity is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2021 solidity contributors.
+#------------------------------------------------------------------------------
+
+set -eu
+
+repo_root="$(dirname "$0")/.."
+robots_txt_path="${repo_root}/docs/_static/robots.txt"
+version_from_script=$("${repo_root}"/scripts/get_version.sh | tr -d '\n')
+last_line_robots=$(tail -n 1 "${robots_txt_path}")
```

Comment:
https://github.com/ethereum/solidity/pull/11206#discussion_r611790475 has not been addressed yet.

Reply:
I actually mentioned about this in here https://github.com/ethereum/solidity/pull/11206#pullrequestreview-638311644 that as of now the script assumes we need to change the last line.

Reply:
Yeah, I missed your answer because you marked the comment as "resolved" (and it was also "minimized" for some reason, making it even harder to find). If a particular point still requires discussion, you should leave it as unresolved so that the reviewer can actually can see it :)

Anyway, I think that the approach based only on detecting `Disallow` is a bit too brittle. I suggested a more reliable one in my original comment there.

Reply:
There's also another possibility that I mentioned in the issue (and that @axic brought up today):

> I think these generator scripts may be much nicer if we didn't try to change/update files in place, rather have a template file which is just extended.
>
> I looked into the issue of generating the binary list for binaries.soliditylang.org and I included that having `README.md.template` which is used in the script would be the easiest.
>
> Perhaps the same applies here too.

I think that would actually be a nicer approach. The main downside is that the script would have to get a list of versions from somewhere. I originally mentioned getting tags from github API but just listing tags using `git` would work too.

@axic Maybe we should really discuss this before @cxxboy gets to deep into rewriting the script.

Reply:
@cameel 
So, just to get a better understanding before I go ahead. According to your suggession, Is this how it's supposed to work?
Our `robots.txt` file will now looks like this,
```
# Allow the latest release
#Disallow: /en/v0.8.3/
# LATEST RELEASE: 0.8.3
```
and when we execute the `update_robots.sh` script, we check the `get_version.sh`'s output against `# LATEST RELEASE` and update the **one line ahead**(like I guess we can have this assumption right?, like the `#Disallow` directive is `N-1`th line if `# LATEST RELEASE` is on the line `N`?) where we have `#Disallow` directive and if the version doesn't match(between `# LATEST VERSION` and `get_version.sh` script) we just remove the `#` and have the updated script like this:
 ```
Disallow: Previously_In_LATEST_RELEASE
#Disallow: New_Version_From_Script
# LATEST RELEASE: New_Version_From_Script
```
right? If if the version does match we don't touch the `robots.txt`
So I guess, I need to now manually add that `# LATEST RELEASE` line and do the rewrite.

Reply:
This is start to look a bit like a bumpversion configuration. Here's an example project we use: https://github.com/ethereum/evmc/blob/master/.bumpversion.cfg

Reply:
Can `bumpversion` handle more than simple search and replace of version string? Here the situation is a tiny bit more complex, i.e. there's one line per version and new lines need to be inserted under certain circumstances.

Personally, I'd prefer to just have a template. Updating the file in place works too but it's messier. E.g. it does not work that well if you skip a version by accident (like it has already happened with 0.8.4).

Reply:
@cameel 
Can you explain a bit about what "template" you're talking about? I don't really understand what you mean when you say we build `robots.txt` from a "template".

Do you mean we have something like `lex` in unix where our script "scans" that "template" file which looks like this
```
"Sitemap:" <SPACE> <URL>
"Disallow:" <SPACE> (<BRANCH> | "en"/<PREVIOUS SOL VERSIONS>/)
```
where in this case, things between `"` is interpreted as a "string"(like it's put as it as) and things between `<` and `>` are like place holders which will be interpreted and filled by our script. For this we need to grab the array of versions from some web endpoint and use it when we interpret this "template"

Is this what you mean by using a "template" to generate `robots.txt` instead of editing files in-place? If this is what you mean, using bash is very limited atleast for me. I need something like Perl (as I don't know Python or Javascript)
(and Perl comes default with most unix-like systems too). I can implement something like this if this is something we need to go with.

Reply:
Kinda, but I would not overcomplicate it with `lex`. The idea would be more like:

`robots.txt.template`:
```robots.txt
Sitemap: http://docs.soliditylang.org/sitemap.xml

# Prevent documentation for the development branches from showing up in search results.
Disallow: /en/develop/
Disallow: /en/breaking/

# Prevent documentation for the older Solidity versions from showing up in search results.
{{ allowed_and_disallowed_versions }}
```

You would just find the `{{ allowed_and_disallowed_versions }}` placeholder and replace it with the `Allow`/`Disallow` list constructed from git tags.

But looks like we can make it even simpler than that. @christianparpart has chimed in with a very nice idea: https://github.com/ethereum/solidity/pull/11317#discussion_r620435405. If it works, it would make the whole system even simpler. Just allow `v0.7.6`, the current version and a few other paths that never change.

This way we could even use `bumpversion` suggested by @axic, though I'd leave this for another PR.

> It will be nice if I have Perl (as I don't know Python or Javascript)

The problem with Perl is really that it's the opposite for us. Most of the team either knows or at least can get by with Python,  JavaScript and Bash. Perl not so much. Even those 3 are a bit much and we'd like to avoid introducing new languages that we require everyone to know. We also do not have any infrastructure for Perl while for Bash and Python we have linting and for Python even unit testing.

In this situation Bash seems to be the only common denominator in this PR unfortunately.

Reply:
OK, so to finally make a decision here, my proposal would be to go with what @christianparpart suggested:

- Remove `robots.txt` and replace it with `robots.txt.template`:
    ```robots.txt
    User-Agent: *
    Sitemap: http://docs.soliditylang.org/sitemap.xml

    Allow: /en/latest/
    Allow: /en/v0.7.6/
    Allow: /en/v{{ LATEST_VERSION }}/

    # Prevent documentation for the development branches and older Solidity
    # versions from showing up in search results.
    Disallow: /en/*
    ```
- The script should create `docs/_static/robots.txt` based on that template.
    - `{{ LATEST_VERSION }}` should be replaced with version from `get_latest_version.sh`.
    - Alternatively you could also use a different syntax for the placeholder. E.g. `${LATEST_VERSION}` as suggested by @christianparpart, which would allow you to render it using Bash interpolation and would let us add other variables in the future.
- I'll create an issue to switch to `bumpversion` in the future so let's keep it simple.

@axic, @christianparpart any objections?

Reply:
@cameel This sounds even nice and simple :)